File: frontend/src/services/tickets.ts
import api from "./api"

export interface Ticket {
  id: string
  customer_id: string
  subject: string
  description: string
  priority: string
  status: string
  assigned_to: string | null
  created_at: string
  updated_at: string
}

export interface TicketCreate {
  customer_id: string
  subject: string
  description: string
  priority?: string
}

export interface TicketComment {
  id: string
  ticket_id: string
  user_id: string | null
  comment: string
  is_internal: boolean
  created_at: string
}

export async function listTickets(params?: {
  status?: string
  priority?: string
  customer_id?: string
  skip?: number
  limit?: number
}) {
  const res = await api.get("/tickets", { params })
  return res.data as Ticket[]
}

export async function getTicket(id: string) {
  const res = await api.get(`/tickets/${id}`)
  return res.data as Ticket
}

export async function createTicket(data: TicketCreate) {
  const res = await api.post("/tickets", data)
  return res.data as Ticket
}

export async function updateTicket(id: string, data: { status?: string; priority?: string; assigned_to?: string }) {
  const res = await api.put(`/tickets/${id}`, data)
  return res.data as Ticket
}

export async function getComments(ticketId: string) {
  const res = await api.get(`/tickets/${ticketId}/comments`)
  return res.data as TicketComment[]
}

export async function addComment(ticketId: string, data: { comment: string; is_internal?: boolean }) {
  const res = await api.post(`/tickets/${ticketId}/comments`, data)
  return res.data as TicketComment
}

File: backend/app/api/v1/tickets.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import AdminOnly, AnyStaff, SupportStaff
from app.models.customer import Customer
from app.models.user import User
from app.schemas.ticket import (
    TicketCommentCreate,
    TicketCommentRead,
    TicketCreate,
    TicketRead,
    TicketUpdate,
)
from app.services.ticket_service import TicketService

router = APIRouter(prefix="/tickets", tags=["tickets"])


@router.get("", response_model=list[TicketRead])
async def list_tickets(
    status: str | None = Query(None),
    priority: str | None = Query(None),
    customer_id: uuid.UUID | None = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(AnyStaff),
):
    service = TicketService(db)
    return await service.list(
        user.organization_id,
        status=status,
        priority=priority,
        customer_id=customer_id,
        skip=skip,
        limit=limit,
    )


@router.post("", response_model=TicketRead, status_code=201)
async def create_ticket(
    data: TicketCreate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(SupportStaff),
):
    service = TicketService(db)
    return await service.create(data, user.organization_id)


@router.get("/{ticket_id}", response_model=TicketRead)
async def get_ticket(
    ticket_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(AnyStaff),
):
    service = TicketService(db)
    ticket = await service.get(ticket_id, user.organization_id)
    if not ticket:
        raise HTTPException(status_code=404, detail="Ticket not found")
    return ticket


@router.put("/{ticket_id}", response_model=TicketRead)
async def update_ticket(
    ticket_id: uuid.UUID,
    data: TicketUpdate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(SupportStaff),
):
    service = TicketService(db)
    ticket = await service.update(ticket_id, data, user.organization_id)
    if not ticket:
        raise HTTPException(status_code=404, detail="Ticket not found")
    return ticket


@router.get("/{ticket_id}/comments", response_model=list[TicketCommentRead])
async def list_comments(
    ticket_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(AnyStaff),
):
    service = TicketService(db)
    ticket = await service.get(ticket_id, user.organization_id)
    if not ticket:
        raise HTTPException(status_code=404, detail="Ticket not found")
    is_admin = user.role == "admin"
    return await service.get_comments(ticket_id, organization_id=user.organization_id, include_internal=is_admin)


@router.post("/{ticket_id}/comments", response_model=TicketCommentRead, status_code=201)
async def add_comment(
    ticket_id: uuid.UUID,
    data: TicketCommentCreate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(SupportStaff),
):
    service = TicketService(db)
    ticket = await service.get(ticket_id, user.organization_id)
    if not ticket:
        raise HTTPException(status_code=404, detail="Ticket not found")
    return await service.add_comment(ticket_id, data, organization_id=user.organization_id, user_id=user.id)
